In [66]:
# Import Modules
import numpy as np 
import pandas as pd
from pandas_profiling import ProfileReport
 
import matplotlib.pyplot as plt
%matplotlib qt

import seaborn as sns
sns.set_style('dark')

In [74]:
file_path = 'C:\\Users\\AnandThirurangaruban\\Desktop\\PGPAIML\\Case Study\\'
df = pd.read_csv(file_path+'loan.csv')
#print(df.head(1))

      id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  54734      80364      25000        25000       19080.0572   36 months   

  int_rate  installment grade sub_grade  ... num_tl_90g_dpd_24m  \
0   11.89%        829.1     B        B4  ...                NaN   

  num_tl_op_past_12m pct_tl_nvr_dlq  percent_bc_gt_75 pub_rec_bankruptcies  \
0                NaN            NaN               NaN                  0.0   

  tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit  \
0       0.0             NaN               NaN            NaN   

  total_il_high_credit_limit  
0                        NaN  

[1 rows x 111 columns]


C:\Users\AnandThirurangaruban\AppData\Local\Temp\ipykernel_34832\2850910134.py:2: DtypeWarning: Columns (47) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path+'loan.csv')


### Dataset Observation

#### Observation: 
####  The dataset has 39717 rows and 111 variables 
####  The dataset has 13 integer, 74 float and  24 categorical variiables.

In [69]:
# Print Column count 
print("Columns: ", len(df.columns))

# Print row count 
print("Rows: ", len(df))

print("Columns  and datatype:"  , df.dtypes)


Columns:  111
Rows:  39717
Columns  and datatype: id                              int64
member_id                       int64
loan_amnt                       int64
funded_amnt                     int64
funded_amnt_inv               float64
                               ...   
tax_liens                     float64
tot_hi_cred_lim               float64
total_bal_ex_mort             float64
total_bc_limit                float64
total_il_high_credit_limit    float64
Length: 111, dtype: object


#### Observation:There are no duplicate and empty rows in loan dataset

In [70]:
# sum it up to check how many columns have all missing values
print('Number of empty Columns:')
print(df.isnull().all(axis=0).sum())

empty_column = df.isnull().all(axis=0)

# Filter DataFrame based on values in the second column (index 1)
print(empty_column)

Number of empty Columns:
54
id                            False
member_id                     False
loan_amnt                     False
funded_amnt                   False
funded_amnt_inv               False
                              ...  
tax_liens                     False
tot_hi_cred_lim                True
total_bal_ex_mort              True
total_bc_limit                 True
total_il_high_credit_limit     True
Length: 111, dtype: bool


In [71]:
print('Number of empty Rows:')
print(df.isnull().all(axis=1).sum())


# Data Acquisition and Preprocessing

#Check the values in df
print("Null Count: ", df.isnull().sum())

print("Shape : ", df.shape)

print("Count of Null:" , df.notna().sum())

print("Missing VAlues:" , df.isna().sum())


Number of empty Rows:
0
Null Count:  id                                0
member_id                         0
loan_amnt                         0
funded_amnt                       0
funded_amnt_inv                   0
                              ...  
tax_liens                        39
tot_hi_cred_lim               39717
total_bal_ex_mort             39717
total_bc_limit                39717
total_il_high_credit_limit    39717
Length: 111, dtype: int64
Shape :  (39717, 111)
Count of Null: id                            39717
member_id                     39717
loan_amnt                     39717
funded_amnt                   39717
funded_amnt_inv               39717
                              ...  
tax_liens                     39678
tot_hi_cred_lim                   0
total_bal_ex_mort                 0
total_bc_limit                    0
total_il_high_credit_limit        0
Length: 111, dtype: int64
Missing VAlues: id                                0
member_id                     

## EDA - Pandas Profiling

In [ ]:
profile = ProfileReport(df.head(1000))
profile.to_file(file_path+ 'loan.html')

### Display Values in each Object

In [ ]:
# How many values available in each object('categorical')
def df_obj_count(df ,feature_name):    
    if df[feature_name].dtype ==  'object':
        print("Column Name :" , df[feature_name].value_counts())

for obj in df.columns:
    if df[obj].dtype == 'object':
        df_obj_count(df, str(obj))

In [ ]:
# Remove columns where object is NA
df_removed = df.drop(['mths_since_last_record','next_pymnt_d','id','pymnt_plan','initial_list_status','collections_12_mths_ex_med','policy_code','application_type','chargeoff_within_12_mths','tax_liens','url','mths_since_last_major_derog','annual_inc_joint','dti_joint','verification_status_joint','acc_now_delinq','tot_coll_amt','tot_cur_bal','open_acc_6m','open_il_6m','open_il_12m','open_il_24m','mths_since_rcnt_il','total_bal_il','il_util','open_rv_12m','open_rv_24m','max_bal_bc','all_util','total_rev_hi_lim','inq_fi','total_cu_tl','inq_last_12m','acc_open_past_24mths','avg_cur_bal','bc_open_to_buy','bc_util','delinq_amnt','mo_sin_old_il_acct','mo_sin_old_rev_tl_op','mo_sin_rcnt_rev_tl_op','mo_sin_rcnt_tl','mort_acc','mths_since_recent_bc','mths_since_recent_bc_dlq','mths_since_recent_inq','mths_since_recent_revol_delinq','num_accts_ever_120_pd','num_actv_bc_tl','num_actv_rev_tl','num_bc_sats','num_bc_tl','num_il_tl','num_op_rev_tl','num_rev_accts','num_rev_tl_bal_gt_0','num_sats','num_tl_120dpd_2m','num_tl_30dpd','num_tl_90g_dpd_24m','num_tl_op_past_12m','pct_tl_nvr_dlq','percent_bc_gt_75','tot_hi_cred_lim','total_bal_ex_mort','total_bc_limit','total_bc_limit','total_il_high_credit_limit'], axis=1)

#df_removed.fillna(df_removed.mean(), inplace=True)

#df_removed.to_csv('C:\\Users\\AnandThirurangaruban\\Desktop\\PGPAIML\\Case Study\\cleaneddata.csv' , index=False)

# df.drop(columns =['C', 'D'])

In [ ]:
# To observe it manually , i'm loading the data in csv
df_removed.to_csv('cleaneddata.csv' , index=False)

In [ ]:
# Data Acquisition and Preprocessing

print("Column Name :" , df_removed.columns)

print("Null Count: ", df_removed.isnull().sum())

print("Rows: ", len(df_removed))

print("Columns: ", len(df_removed.columns))

print("Shape : ", df_removed.shape)

print("Info :"  ,df_removed.info())

print("Count of Null:" ,df_removed.isna().sum()) 

print("Count of Null:" , df_removed.notna().sum())

print("Missing VAlues:" , df_removed.isna().sum())


In [ ]:
# Check purpose of loan in total loans issued
(df.purpose.value_counts()*100)/len(df)

## <u>Target Variable</u>  [Target variable is the main focus of a statistical analysis or machine learning model]
### The target variable is <font color='Red'>loan_Status</font>
The objective is to determine how data is used to minimise the risk of losing money while lending to customers. Hence this particular variable <font color='red'>'loan_status' </font> is the target variable.

In [ ]:
#pulling out object dtype columns to see which columns require transformation
df.select_dtypes(object).head()

#df.select_dtypes(int).head()

Columns such as <font color='red'>term, int_rate, emp_length</font> must be converted to <font color='green'>float type</font>. Also we need to convert <font color='red'>issue_d</font> from object type to <font color='green'>date type</font>.

In [ ]:
#first removing the unwanted character from the string, then casting the string object as float
df_removed['term']=df_removed['term'].replace('[(<)a-zA-Z(%+)-]','',regex=True).astype(float)
df_removed['int_rate']=df_removed['int_rate'].replace('[(<)a-zA-Z(%+)-]','',regex=True).astype(float)
df_removed['emp_length']=df_removed['emp_length'].replace('[(<)a-zA-Z(%+)-]','',regex=True).astype(float)

### Univariate Analysis - Numerical Variables

In [ ]:
plt.figure(figsize=(18, 6))
plt.subplot(1, 2, 1)

# Convert infinite values to NaN before operating
with pd.option_context('mode.use_inf_as_na', True):
    # Your operations on the DataFrame here
    df_removed.dropna(inplace=True)  # For example, drop rows with NaN values

df_removed['annual_inc'].hist(bins=10)
plt.title("Loan Amount ")

plt.subplot(1, 2, 2)
plt.grid()
plt.hist(df_removed['loan_amnt'])
plt.title("Log Loan Amount ")

plt.show()  

# Lets see distribution of three loan amount fields using distribution plot.
plt.figure(figsize=(15,8),facecolor='b')
sns.set_style("dark")
# subplot 1
plt.subplot(2, 3, 1)


# Custom color palette
colors = ['red', 'green', 'blue', 'orange', 'purple', 'cyan']

# Plot histogram with different colors

#plt.show()


viz = sns.histplot(df['loan_amnt'], color='lightblue', edgecolor='black',palette=colors, bins=10,kde=True)  # Use a custom color palette,kde=True)
viz.set_title('Loan Amount - Distribution Plot',fontsize=14,color='w')
viz.set_xlabel('Loan Amount',fontsize=14,color='w')
viz.set_facecolor('black') 


# subplot 2
plt.subplot(2, 3, 2)
viz = sns.histplot(df['funded_amnt'], color='lightblue', edgecolor='black',palette=colors, bins=10,kde=True) 
viz.set_title('Fund Amount - Distribution Plot',fontsize=14,color='w')
viz.set_xlabel('Fund Amount',fontsize=14,color='w')
viz.set_facecolor('black') 

# subplot 2
plt.subplot(2, 3, 3)
viz = sns.histplot(df['funded_amnt_inv'], color='lightblue', edgecolor='black',palette=colors, bins=10,kde=True) 
viz.set_title('Fund Amount Inv. - Distribution Plot',fontsize=14,color='w')
viz.set_xlabel('Fund Amount Inv.',fontsize=14,color='w')
viz.set_facecolor('black') 
plt.show()

In [81]:
plt.figure(figsize=(14,8),facecolor='b')
sns.set_style("dark")
plot = sns.countplot(y="purpose",data=df_removed,hue='loan_status',palette='GnBu_d')
plot.set_title('Purpose Of Loan',fontsize=14,color='w')
plot.set_ylabel('Purpose Of Loan',fontsize=14,color = 'w')
plot.set_xlabel('Loan Application Count',fontsize=14,color = 'w')      
plt.show()

In [90]:
df_removed.drop(columns=['funded_amnt','funded_amnt_inv'],inplace=True)

df_removed.loan_status=df.loan_status.apply(lambda x: 0 if x[0]=='F' else 1)



In [91]:
#here we're plotting bar graph for loan_status and grade
plt.title('loan_status vs grade \n', fontdict={'fontsize':20,'color':'r'})
sns.set_style('darkgrid')
sns.barplot(x=df_removed['grade'].sort_values(),y=df_removed['loan_status'])
plt.show()

In [92]:
#here we're plotting bar graph for loan_status and sub_grade
plt.title('loan_status vs sub_grade \n', fontdict={'fontsize':20,'color':'r'})
plt.xticks(rotation=90)
sns.barplot(x=df_removed['sub_grade'].sort_values(),y=df_removed['loan_status'])
plt.show()

TypeError: Neither the `x` nor `y` variable appears to be numeric.